Code/GraphMol/MolDraw2D/DrawShape.cpp

```diff
@@ -0,0 +1,552 @@
+//
+//  Copyright (C) 2021-2022 David Cosgrove and other RDKit contributors
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+// Original author: David Cosgrove (CozChemIx Limited)
+//
+
+#include <GraphMol/MolDraw2D/MolDraw2DDetails.h>
+#include <GraphMol/MolDraw2D/DrawShape.h>
+#include <GraphMol/MolDraw2D/DrawText.h>
+#include <GraphMol/MolDraw2D/MolDraw2D.h>
+#include <GraphMol/MolDraw2D/StringRect.h>
+
+namespace RDKit {
+namespace MolDraw2D_detail {
+
+// ****************************************************************************
+DrawShape::DrawShape(const std::vector<Point2D> &points, int lineWidth,
+                     bool scaleLineWidth, DrawColour lineColour, bool fill,
+                     int atom1, int atom2, int bond)
+    : points_(points),
+      lineWidth_(lineWidth),
+      scaleLineWidth_(scaleLineWidth),
+      lineColour_(lineColour),
+      fill_(fill),
+      atom1_(atom1),
+      atom2_(atom2),
+      bond_(bond) {}
+
+// ****************************************************************************
+void DrawShape::draw(MolDraw2D &drawer) {
+  // the various myDraw functions may over-ride these.
+  const auto ocolour = drawer.colour();
+  drawer.setColour(lineColour_);
+  const auto olw = drawer.lineWidth();
+  drawer.setLineWidth(lineWidth_);
+  const auto ofill = drawer.fillPolys();
+  drawer.setFillPolys(fill_);
+  const auto lineScale = drawer.drawOptions().scaleBondWidth;
+  drawer.drawOptions().scaleBondWidth = scaleLineWidth_;
+  drawer.setActiveAtmIdx(atom1_, atom2_);
+  drawer.setActiveBndIdx(bond_);
+
+  myDraw(drawer);
+
+  drawer.setActiveAtmIdx();
+  drawer.setActiveBndIdx();
+  drawer.setColour(ocolour);
+  drawer.setLineWidth(olw);
+  drawer.setFillPolys(ofill);
+  drawer.drawOptions().scaleBondWidth = lineScale;
+}
+
+// ****************************************************************************
+void DrawShape::findExtremes(double &xmin, double &xmax, double &ymin,
+                             double &ymax) const {
+  for (auto &p : points_) {
+    xmin = std::min(xmin, p.x);
+    xmax = std::max(xmax, p.x);
+    ymin = std::min(ymin, p.y);
+    ymax = std::max(ymax, p.y);
+  }
+}
+
+// ****************************************************************************
+void DrawShape::scale(const Point2D &scale_factor) {
+  for (auto &p : points_) {
+    p.x *= scale_factor.x;
+    p.y *= scale_factor.y;
+  }
+}
+
+// ****************************************************************************
+void DrawShape::move(const Point2D &trans) {
+  for (auto &p : points_) {
+    p += trans;
+  }
+}
+
+// ****************************************************************************
+bool DrawShape::doesRectClash(const StringRect &rect, double padding) const {
```

Comment:
This will remove a compiler warning:
```suggestion
bool DrawShape::doesRectClash(const StringRect &, double ) const {
```